File: ios/Connect/ConnectTests/MockModels/MockConnectionsRepo.swift
//
//  MockConnectionsRepo.swift
//  Quartier
//
//  Created by Emile Turcotte on 2025-07-23.
//

import Testing
@testable import Quartier

class MockConnectionsRepo : ConnectionsRepository {
    private(set) var connections = [
        Connection(requested_by: "0", requested_for: "1", status: "accepted"),
        Connection(requested_by: "2", requested_for: "0", status: "accepted"),
        Connection(requested_by: "0", requested_for: "3", status: "pending"),
        Connection(requested_by: "4", requested_for: "0", status: "pending"),
        Connection(requested_by: "0", requested_for: "5", status: "invalid_status")
    ]
    
    func getConnections() async throws -> [Connection] {
        return connections
    }
    
    func getConnectionWithUser(requestedID: String) async throws -> Connection? {
        return connections.first(where: {
            ($0.requested_by == "0" && $0.requested_for == requestedID) ||
            ($0.requested_by == requestedID && $0.requested_for == "0")
        })
    }
    
    func requestConnection(requestedID: String) async throws -> ConnectionResult? {
        if(requestedID == "0") { return .cannotConnectWithSelf }
        connections.append(Connection(requested_by: "0", requested_for: requestedID, status: "pending"))
        return .requested
    }
    
    func acceptConnection(requestedID: String) async throws {
        let nullableIndex = connections.firstIndex(where: {
            $0.requested_by == requestedID && $0.requested_for == "0"
        })
        if let index = nullableIndex {
            connections[index] = Connection(requested_by: requestedID, requested_for: "0", status: "accepted")
        }
    }
    
    func deleteConnection(requestedID: String) async throws {
        connections.removeAll {
            $0.requested_by == requestedID && $0.requested_for == "0" && $0.status == "pending"
        }
    }
}

File: android/app/src/test/java/com/quartier/quartier/mock_models/MockConnectionsRepo.kt
package com.quartier.quartier.mock_models

import com.quartier.quartier.database.Connection
import com.quartier.quartier.database.ConnectionRequestResult
import com.quartier.quartier.database.ConnectionsRepository
import com.quartier.quartier.database.User
import com.quartier.quartier.database.UserRepository

//Contains fake connections, test behavior for all of them, imitates the Connections table in DB

class MockConnectionsRepo() : ConnectionsRepository {
    private val connections: MutableList<Connection> = mutableListOf(
        Connection("0", "1", "accepted"),
        Connection("2", "0", "accepted"),
        Connection("0", "3", "pending"),
        Connection("4", "0", "pending"),
        Connection("0", "5", "invalid_status"))

    override suspend fun getConnections(): List<Connection> {return connections.toList()}

    override suspend fun getConnectionWithUser(requestedId: String): Connection? {
        return connections.find {
            (it.requested_by == "0" && it.requested_for == requestedId) || (it.requested_by == requestedId && it.requested_for == "0")
        }
    }

    override suspend fun connectionsToUsers(connections: List<Connection>, userRepository: UserRepository) : List<User> {
        val uid = "0"

        val connectionsMap = connections.associateBy(
            {if (it.requested_by == uid) it.requested_for else it.requested_by}, //find the other user's id
            {it.status}
        )
        val users = userRepository.getUsers(connectionsMap.keys.toList())
        return users.map { user ->
            User(user.id, user.name, user.job, user.pfp_url, connectionsMap[user.id])
        }
    }

    override suspend fun requestConnection(requestedId: String) : ConnectionRequestResult {
        if(requestedId == "0") return ConnectionRequestResult.CannotConnectWithSelf
        connections.add(Connection("0", requestedId, "pending"))
        return ConnectionRequestResult.Success
    }

    override suspend fun acceptConnection(user2Id: String) {
        val index = connections.indexOfFirst {
            it.requested_by == user2Id && it.requested_for == "0"
        }
        if(index != -1) {
            connections[index] = Connection(user2Id, "0", "accepted")
        }
    }

    override suspend fun deleteConnection(user2Id: String) {
        connections.remove(Connection(user2Id, "0", "pending"))
    }
}